_posts/2019-11-19-demo-fraud-detection.md

```diff
@@ -0,0 +1,220 @@
+
+#### How to run:
+
+ Execute the following commands to build the project from sources and start all required services, including Apache Flink and Apache Kafka clusters.
+
+```bash
+git clone https://github.com/afedulov/fraud-detection-demo
+docker build -t demo-fraud-webapp:latest -f webapp/webapp.Dockerfile webapp/
+docker build -t flink-job-fraud-demo:latest -f flink-job/Dockerfile flink-job/
+docker-compose -f docker-compose-local-job.yaml up
+```
+
+__Note__: Dependencies are stored in a cached Docker layer. If you later only modify the source code, not the dependencies, you can expect significantly shorter packaging times for the subsequent builds.
+
+When all components are up and running, go to `localhost:5656` in your browser.
+
+__Note__: you might need to change exposed ports in _docker-compose-local-job.yaml_ in case of collisions.
+
+The demo comes with a set of predefined rules. You can simply click the _Start_ button and after some time you should observe alerts displayed on the right side of the screen. Those alerts are the results of Flink evaluating the generated transactions stream against the predefined rules.
+
+ The sample application consists of three main parts:  
+
+ 1. Fraud Detection application (Apache Flink)  
+ 1. Frontend (React)  
+ 1. Backend (SpringBoot)  
+
+
+ Backend includes a Transactions Generator, which sends emulated transactions to Flink via Kafka. Whenever a new rule definition is created in the UI, the backend also pushes it to Flink via a separate topic. Alerts generated by Flink are consumed by the Backend and relayed to the UI via WebSockets.
+
+ <center>
+ <img src="{{ site.baseurl }}/img/blog/2019-11-19-demo-fraud-detection/architecture.png" width="800px" alt="Figure 2: Demo Components"/>
+ <br/>
+ <i><small>Figure 2: Demo Components</small></i>
+ </center>
+ <br/>
+
+
+### Dynamic Data Partitioning
+
+Let us start with formulating a sample rule definition for our fraud detection system as a functional requirement:  
+
+"Whenever the **sum** of accumulated **payment amount** from the same **beneficiary** to the same **payee** within the **duration of a week** is **greater** than **1 000 000 $** - fire an alert."
+
+From this formulation we can extract the following parameters that we would like to be able to specify in a system which allows flexibility in rules definition:  
+
+1. Aggregation field (payment amount)  
+1. Grouping fields (beneficiary + payee)  
+1. Aggregation function (sum)  
+1. Window duration (1 week)  
+1. Limit (1 000 000)  
+1. Limit operator (greater)  
+
+Accordingly, we will use the following simple JSON format to define the aforementioned parameters.
+
+Examples JSON:
+
+```json  
+{
+  "ruleId": 1,
+  "ruleState": "ACTIVE",
+  "groupingKeyNames": ["beneficiaryId", "payeeId"],
+  "aggregateFieldName": "paymentAmount",
+  "aggregatorFunctionType": "SUM",
+  "limitOperatorType": "GREATER",
+  "limit": 1000000,
+  "windowMinutes": 10080
+}
+```
```

Comment:
Hmmm, good point. It could be better placed in the second part in the series describing dynamic rules updates - there it makes sense to explain that it is pretty straightforward to pack required updates into a simple format and send over a Kafka topic to Flink.

Reply:
I've decided that I want to transition from abstract idea of the parameters to something concrete here, so I will keep it.